File: client-core/src/main/java/hivens/core/api/SmartyNetworkService.java
package hivens.core.api;

import com.google.gson.Gson;
import hivens.core.api.dto.SmartyResponse;
import hivens.core.api.dto.SmartyServer;

import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.security.MessageDigest;
import java.util.Collections;
import java.util.List;

public class SmartyNetworkService {

    private static final String API_URL = "https://www.smartycraft.ru/launcher2/index.php";
    private static final String OFFICIAL_JAR_URL = "https://www.smartycraft.ru/downloads/smartycraft.jar";
    private static final String CACHED_HASH_FILE = "smarty_hash.cache";

    // Дефолтный хеш (можно обновлять вручную при релизах)
    private static String currentHash = "5515a4bdd5f532faf0db61b8263d1952";

    // Прокси настройки
    private static final Proxy SMARTY_PROXY = new Proxy(Proxy.Type.SOCKS, new InetSocketAddress("proxy.smartycraft.ru", 1080));

    static {
        // Установка аутентификатора для URLConnection
        Authenticator.setDefault(new Authenticator() {
            @Override
            protected PasswordAuthentication getPasswordAuthentication() {
                return new PasswordAuthentication("proxyuser", "proxyuserproxyuser".toCharArray());
            }
        });
    }

    public SmartyNetworkService() {
        // Пытаемся загрузить кэшированный хеш с диска
        File cache = new File(CACHED_HASH_FILE);
        if (cache.exists()) {
            try {
                currentHash = Files.readString(cache.toPath()).trim();
            } catch (IOException ignored) {}
        }
    }

    public List<SmartyServer> getServers() {
        try {
            // Попытка 1: С текущим хешем
            List<SmartyServer> servers = tryGetServers(currentHash);

            if (servers == null) {
                System.out.println("Сервер требует обновления хеша...");
                String newHash = downloadAndCalculateHash();
                if (newHash != null) {
                    currentHash = newHash;
                    // Сохраняем новый хеш на будущее
                    try {
                        Files.writeString(Path.of(CACHED_HASH_FILE), newHash);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                    servers = tryGetServers(newHash);
                }
            }
            return servers != null ? servers : Collections.emptyList();
        } catch (Exception e) {
            e.printStackTrace();
            return Collections.emptyList();
        }
    }

    private List<SmartyServer> tryGetServers(String hash) throws IOException {
        String jsonBody = "{" +
                "\"version\": \"3.6.2\"," + // Версия которую мы эмулируем
                "\"cheksum\": \"" + hash + "\"," +
                "\"format\": \"jar\"," +
                "\"testModeKey\": \"false\"," +
                "\"debug\": \"false\"" +
                "}";

        String responseJson = sendPostRequest("loader", jsonBody);

        // Очистка ответа
        int jsonStart = responseJson.indexOf("{");
        if (jsonStart != -1) {
            responseJson = responseJson.substring(jsonStart);
        }

        try {
            Gson gson = new Gson();
            SmartyResponse response = gson.fromJson(responseJson, SmartyResponse.class);

            if ("UPDATE".equals(response.status)) {
                return null; // Триггер для пересчета хеша
            }
            if ("OK".equals(response.status)) {
                return response.servers;
            }
        } catch (Exception e) {
            System.err.println("Ошибка парсинга: " + responseJson);
        }
        return Collections.emptyList();
    }

    private String downloadAndCalculateHash() {
        try {
            Path tempJar = Files.createTempFile("smarty_temp", ".jar");
            System.out.println("Скачивание официального лаунчера для обхода защиты...");

            URL url = new URL(OFFICIAL_JAR_URL);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection(SMARTY_PROXY);
            conn.setConnectTimeout(15000);
            conn.setReadTimeout(60000);

            try (InputStream in = conn.getInputStream()) {
                Files.copy(in, tempJar, StandardCopyOption.REPLACE_EXISTING);
            }

            MessageDigest digest = MessageDigest.getInstance("MD5");
            try (InputStream fis = Files.newInputStream(tempJar)) {
                byte[] byteArray = new byte[1024];
                int bytesCount;
                while ((bytesCount = fis.read(byteArray)) != -1) {
                    digest.update(byteArray, 0, bytesCount);
                }
            }
            Files.deleteIfExists(tempJar);

            byte[] bytes = digest.digest();
            StringBuilder sb = new StringBuilder();
            for (byte aByte : bytes) {
                sb.append(Integer.toString((aByte & 0xff) + 0x100, 16).substring(1));
            }
            return sb.toString();

        } catch (Exception e) {
            System.err.println("Не удалось получить новый хеш: " + e.getMessage());
            return null;
        }
    }

    private String sendPostRequest(String action, String jsonBody) throws IOException {
        URL url = new URL(API_URL);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection(SMARTY_PROXY);
        conn.setRequestMethod("POST");
        conn.setDoOutput(true);
        // Важные заголовки для маскировки
        conn.setRequestProperty("User-Agent", "SMARTYlauncher/3.6.2");
        conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");

        String urlParameters = "action=" + action + "&json=" + URLEncoder.encode(jsonBody, StandardCharsets.UTF_8);

        try (DataOutputStream wr = new DataOutputStream(conn.getOutputStream())) {
            wr.writeBytes(urlParameters);
            wr.flush();
        }

        try (BufferedReader in = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = in.readLine()) != null) response.append(line);
            return response.toString();
        }
    }
}

File: client-core/src/main/java/hivens/core/api/SmartyNetworkService.kt
package hivens.core.api

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import hivens.config.ServiceEndpoints
import hivens.core.api.dto.SmartyResponse
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.asRequestBody
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.net.InetSocketAddress
import java.net.Proxy
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.security.MessageDigest
import java.util.concurrent.TimeUnit

class SmartyNetworkService(baseClient: OkHttpClient, private val gson: Gson) {

    private val logger = LoggerFactory.getLogger(SmartyNetworkService::class.java)

    // URL официального лаунчера для обхода защиты
    private val officialJarUrl = "https://www.smartycraft.ru/downloads/smartycraft.jar"
    private val cachedHashFile = File("smarty_hash.cache")

    // Дефолтный хеш (обновится сам, если устарел)
    private var currentHash = "5515a4bdd5f532faf0db61b8263d1952"

    private val client: OkHttpClient = baseClient.newBuilder()
        .proxy(Proxy(Proxy.Type.SOCKS, InetSocketAddress("proxy.smartycraft.ru", 1080)))
        .proxyAuthenticator { _, response ->
            val credential = Credentials.basic("proxyuser", "proxyuserproxyuser")
            response.request.newBuilder()
                .header("Proxy-Authorization", credential)
                .build()
        }
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    init {
        // Восстанавливаем хеш из кэша при запуске
        if (cachedHashFile.exists()) {
            try {
                currentHash = Files.readString(cachedHashFile.toPath()).trim()
                logger.info("Loaded cached hash: $currentHash")
            } catch (e: IOException) {
                logger.warn("Failed to read hash cache", e)
            }
        }
    }

    /**
     * Получает данные для Дашборда (Серверы + Новости).
     * Если хеш устарел (статус UPDATE), автоматически обновляет его.
     */
    fun getDashboardResponse(): SmartyResponse {
        var response = tryGetDashboard(currentHash)

        if (response.status == "UPDATE") {
            logger.info("Protection triggered (Status: UPDATE). Initiating bypass...")

            val newHash = downloadAndCalculateHash()

            if (newHash != null) {
                logger.info("Bypass successful. New hash: $newHash")
                currentHash = newHash
                saveHashToCache(newHash)

                // Повторяем запрос с новым хешем
                response = tryGetDashboard(newHash)
            } else {
                logger.error("Bypass failed. Could not calculate new hash.")
            }
        }

        return response
    }

    /**
     * Загружает скин или плащ на сервер.
     */
    fun uploadAsset(file: File, type: String, session: String): String {
        // Используем современные расширения OkHttp (избавляемся от Deprecated)
        val mediaType = "image/png".toMediaTypeOrNull()
        val fileBody = file.asRequestBody(mediaType)

        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("action", "upload")
            .addFormDataPart("type", type)
            .addFormDataPart("session", session)
            .addFormDataPart("file", file.name, fileBody)
            .build()

        val request = Request.Builder()
            .url(ServiceEndpoints.AUTH_LOGIN)
            .post(requestBody)
            .header("User-Agent", "SMARTYlauncher/3.6.2")
            .build()

        return try {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return "Ошибка сети: ${response.code}"
                response.body?.string()?.trim() ?: "Ошибка: Пустой ответ"
            }
        } catch (e: Exception) {
            logger.error("Upload asset failed", e)
            "Ошибка: ${e.message}"
        }
    }

    // --- Приватные методы ---

    private fun tryGetDashboard(hash: String): SmartyResponse {
        val jsonBody = """
            {
                "version": "3.6.2",
                "cheksum": "$hash",
                "format": "jar",
                "testModeKey": "false",
                "debug": "false"
            }
        """.trimIndent()

        val formBody = FormBody.Builder()
            .add("action", "loader")
            .add("json", jsonBody)
            .build()

        val request = Request.Builder()
            .url(ServiceEndpoints.AUTH_LOGIN)
            .post(formBody)
            .header("User-Agent", "SMARTYlauncher/3.6.2")
            .build()

        try {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return SmartyResponse()

                var rawJson = response.body?.string() ?: return SmartyResponse()

                // Очистка от HTML-мусора
                val jsonStart = rawJson.indexOf("{")
                if (jsonStart != -1) {
                    rawJson = rawJson.substring(jsonStart)
                }

                return try {
                    gson.fromJson(rawJson, SmartyResponse::class.java)
                } catch (e: JsonSyntaxException) {
                    logger.error("Parsing error: $rawJson", e)
                    SmartyResponse()
                }
            }
        } catch (e: Exception) {
            logger.error("Network error in tryGetDashboard", e)
            return SmartyResponse()
        }
    }

    private fun downloadAndCalculateHash(): String? {
        try {
            logger.info("Downloading official launcher form $officialJarUrl...")

            val tempJar = Files.createTempFile("smarty_temp", ".jar")

            val request = Request.Builder()
                .url(officialJarUrl)
                .header("User-Agent", "SMARTYlauncher/3.6.2")
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    logger.error("Download failed: HTTP ${response.code}")
                    return null
                }
                response.body?.byteStream()?.use { input ->
                    Files.copy(input, tempJar, StandardCopyOption.REPLACE_EXISTING)
                }
            }

            logger.info("Calculating MD5 hash...")
            val digest = MessageDigest.getInstance("MD5")
            Files.newInputStream(tempJar).use { fis ->
                val buffer = ByteArray(1024)
                var bytesCount: Int
                while (fis.read(buffer).also { bytesCount = it } != -1) {
                    digest.update(buffer, 0, bytesCount)
                }
            }

            Files.deleteIfExists(tempJar)

            val bytes = digest.digest()
            val sb = StringBuilder()
            for (b in bytes) {
                sb.append(((b.toInt() and 0xff) + 0x100).toString(16).substring(1))
            }
            return sb.toString()

        } catch (e: Exception) {
            logger.error("Failed to calculate new hash", e)
            return null
        }
    }

    private fun saveHashToCache(hash: String) {
        try {
            Files.writeString(cachedHashFile.toPath(), hash)
            logger.info("New hash saved to cache.")
        } catch (e: IOException) {
            logger.error("Failed to save hash cache", e)
        }
    }
}
